src/qibolab/instruments/rfsoc/driver.py

```diff
@@ -498,14 +510,38 @@ def convert_sweep_results(
                     qubit = qubits[ro_pulse.qubit]
                     discriminated_shots = self.classify_shots(i_vals, q_vals, qubit)
                     if execution_parameters.averaging_mode is AveragingMode.CYCLIC:
-                        discriminated_shots = np.mean(discriminated_shots, axis=0)
-                    result = execution_parameters.results_type(discriminated_shots)
+                        _, counts = np.unique(
+                            discriminated_shots, return_counts=True, axis=0
+                        )
+                        freqs = counts / np.shape(discriminated_shots)[0]
+                        result = execution_parameters.results_type(
+                            freqs, discriminated_shots
+                        )
+                    else:
+                        result = execution_parameters.results_type(discriminated_shots)
+
                 else:
                     result = execution_parameters.results_type(i_vals + 1j * q_vals)
 
                 results[ro_pulse.qubit] = results[ro_pulse.serial] = result
         return results
 
+    @staticmethod
+    def reshape_sweep_results(results, sweepers, execution_parameters):
+        shape = [len(sweeper.values) for sweeper in sweepers]
+        if execution_parameters.averaging_mode is not AveragingMode.CYCLIC:
+            shape.insert(0, execution_parameters.nshots)
+        new_res = {}
+        for key, value in results.items():
+            if isinstance(value, IntegratedResults):
+                data = value.voltage
+            elif isinstance(value, AveragedSampleResults):
+                data = value.statistical_frequency
+            else:
+                data = value.samples
+            new_res[key] = type(value)(data.reshape(shape))
+        return new_res
```

Comment:
Slightly more concise:

```suggestion
        def data(value):
            if isinstance(value, IntegratedResults):
                data = value.voltage
            elif isinstance(value, AveragedSampleResults):
                data = value.statistical_frequency
            else:
                data = value.samples
            return type(value)(data.reshape(shape))
        return {key: data(value) for key, value in results.items()}
```

It could be much better if there were a `.data` property in the various results (pointing to the correct one for each type). Or even better if they were just NumPy arrays...